src/grpc/hpr_packet_up.erl

```diff
@@ -119,6 +128,37 @@ to_map(PacketRecord) ->
         signature => PacketRecord#packet_router_packet_up_v1_pb.signature
     }.
 
+-spec type(Packet :: packet()) -> packet_type().
+type(Packet) ->
+    case ?MODULE:payload(Packet) of
+        <<?JOIN_REQUEST:3, _:5, AppEUI:64/integer-unsigned-little,
+            DevEUI:64/integer-unsigned-little, _DevNonce:2/binary, _MIC:4/binary>> ->
+            {join_req, {AppEUI, DevEUI}};
+        (<<FType:3, _:5, DevAddr:32/integer-unsigned-little, _ADR:1, _ADRACKReq:1, _ACK:1, _RFU:1,
+            FOptsLen:4, _FCnt:16/little-unsigned-integer, _FOpts:FOptsLen/binary,
+            PayloadAndMIC/binary>>) when
+            (FType == ?UNCONFIRMED_UP orelse FType == ?CONFIRMED_UP) andalso
+                %% MIC is 4 bytes, so the binary must be at least that long
+                erlang:byte_size(PayloadAndMIC) >= 4
+        ->
+            Body = binary:part(PayloadAndMIC, {0, byte_size(PayloadAndMIC) - 4}),
```

Comment:
Since we're not looking at any value past the `Port`, I wonder if this would be clearer if we check the length and then pattern match, something like:
```
if
    byte_size(PayloadAndMIC) == 4 -> undefined;
    true -> <<Port:8, _/binary>>, Port
end
```